### PR TITLE
Update kotlin test dependency used by test suites and build init to the latest junit5 version

### LINKED ISSUE #19761: Update DefaultJvmTestSuite and BuildInit to use JUnit 5 when applying useKotlinTest
When I try to configure `JvmTestSuite` (`DefaultJvmTestSuite`) with `useKotlinTest("1.6.10")`
Gradle uses JUnit 4 implementation of `kotlin.test` library: [org.jetbrains.kotlin:kotlin-test-junit:1.6.10](https://mvnrepository.com/artifact/org.jetbrains.kotlin/kotlin-test-junit/1.6.10). See also: [`DefaultJvmTestSuite.java#L53`](https://github.com/gradle/gradle/blob/v7.3.3/subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmTestSuite.java#L53).

When using Kotlin JVM, I expect that `kotlin.test` library with JUnit 5 implementation is included in test compile-time and runtime dependencies: [org.jetbrains.kotlin:kotlin-test-junit5:1.6.10](https://mvnrepository.com/artifact/org.jetbrains.kotlin/kotlin-test-junit5/1.6.10).

```build.gradle.kts
plugins {
    kotlin("jvm") version "1.6.10"
}

group = "ru.pihanya.reproduce.gradle"

repositories {
    mavenCentral()
}

dependencies {
    implementation(kotlin("stdlib"))
}

extensions.configure<org.gradle.testing.base.TestingExtension> {
    suites.withType<org.gradle.api.plugins.jvm.internal.DefaultJvmTestSuite>()
        .configureEach { useKotlinTest("1.6.10") }
}
```

### Expected Behavior

[org.jetbrains.kotlin:kotlin-test-junit5](https://mvnrepository.com/artifact/org.jetbrains.kotlin/kotlin-test-junit5) is used when call `DefaultJvmTestSuite#userKotlinTest`.

### Current Behavior

[org.jetbrains.kotlin:kotlin-test-junit](https://mvnrepository.com/artifact/org.jetbrains.kotlin/kotlin-test-junit) is used when call `DefaultJvmTestSuite#userKotlinTest`.

### Context

Switching to JUnit 5 `kotlin.test` implementation will break backward compatibility. But [Test Suites](https://docs.gradle.org/current/samples/sample_incubating_jvm_multi_project_with_additional_test_types.html) is [incubrating](https://docs.gradle.org/current/userguide/feature_lifecycle.html#sec:incubating_state) feature so there should not be a great problem to break backward compatibility.

**There's also a workaround solution for this problem (see spoiler below).**
<details>
<summary>
<b>Workaround solution</b> (spoiler)
</summary>

```build.gradle.kts
plugins { 
    kotlin("jvm") version "1.6.10"
}

group = "ru.pihanya.reproduce.gradle"

repositories {
    mavenCentral()
}

dependencies {
    implementation(kotlin("stdlib"))
}

extensions.configure<org.gradle.testing.base.TestingExtension> {
    suites.withType<org.gradle.api.plugins.jvm.internal.DefaultJvmTestSuite>().configureEach {
        dependencies {
            implementation("org.jetbrains.kotlin:kotlin-test-junit5:1.6.10")
        }
    }
}
```

</details>

### Steps to Reproduce 

1. Clone repository with reproduction of the issue
```shell
git clone git@github.com:Pihanya/gradle-reproduce.git
```
2. Change the current directory to the cloned repository
```shell
cd gradle-reproduce
```
3. Generate Gradle wrapper for the project:
```shell
gradle wrapper --gradle-version 7.3.3
```
4. Reproduce problem by executing task `checkDefaultJvmTestSuiteUsesJunit5`:
```shell
./gradlew checkDefaultJvmTestSuiteUsesJunit5
```

**Final script:**
```shell
git clone git@github.com:Pihanya/gradle-reproduce.git
cd gradle-reproduce
gradle wrapper --gradle-version 7.3.3
./gradlew checkDefaultJvmTestSuiteUsesJunit5
```

### Your Environment

Gradle: `7.3.3`
Build scan URL: https://gradle.com/s/rvizzusvf6gzw

### PR DESCRIPTION
**Pull request by tresat**
Fixes #19761

### CODE REVIEW

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test QFL

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmTestSuite.java [hunk 1]
```diff
@@ -50,7 +50,7 @@ public enum Frameworks {
         JUNIT4("junit:junit", "4.13"),
         JUNIT_JUPITER("org.junit.jupiter:junit-jupiter", "5.7.2"),
         SPOCK("org.spockframework:spock-core", "2.1-groovy-3.0"),
-        KOTLIN_TEST("org.jetbrains.kotlin:kotlin-test-junit", "1.6.21"),
+        KOTLIN_TEST("org.jetbrains.kotlin:kotlin-test-junit5", "1.6.21"),
         TESTNG("org.testng:testng", "7.4.0"),
```
Comment (big-guy) [inline subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmTestSuite.java:54]: As a separate PR, could you make a pass at updating the versions here?

File: subprojects/build-init/src/main/java/org/gradle/buildinit/plugins/internal/JvmGradlePluginProjectInitDescriptor.java [hunk 2]
```diff
@@ -76,12 +80,12 @@ public void generateProjectBuildScript(String projectName, InitSettings settings
...
                 b.propertyAssignment(null, "testClassesDirs", buildScriptBuilder.propertyExpression(functionalTestSourceSet, "output.classesDirs"), true);
                 b.propertyAssignment(null, "classpath", buildScriptBuilder.propertyExpression(functionalTestSourceSet, "runtimeClasspath"), true);
-                if (getTestFrameworks().contains(BuildInitTestFramework.SPOCK) || getTestFrameworks().contains(BuildInitTestFramework.JUNIT_JUPITER)) {
+                if (!Sets.intersection(FRAMEWORKS_USING_JUNIT_PLATFORM, getTestFrameworks()).isEmpty()) {
                     b.methodInvocation(null, "useJUnitPlatform");
                 }
             });
             buildScriptBuilder.taskMethodInvocation("Run the functional tests as part of `check`", "check", "Task", "dependsOn", functionalTest);
-            if (getTestFrameworks().contains(BuildInitTestFramework.SPOCK) || getTestFrameworks().contains(BuildInitTestFramework.JUNIT_JUPITER)) {
+            if (!Sets.intersection(FRAMEWORKS_USING_JUNIT_PLATFORM, getTestFrameworks()).isEmpty()) {
```
Comment (big-guy) [inline subprojects/build-init/src/main/java/org/gradle/buildinit/plugins/internal/JvmGradlePluginProjectInitDescriptor.java:88]: I think these conditions are always true now. 

What `getTestFrameworks()` returns is weird and it's only used here in the descriptors. The Java, Groovy and Kotlin Gradle plugin descriptors always return a single item and it's one of the frameworks from the set here. 

I would remove these conditionals. We are always going to need `useJUnitPlatform` now.

**Review by big-guy — CHANGES_REQUESTED** [review]

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmTestSuite.java (hunk 1 above)
Comment (tresat) [reply subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmTestSuite.java:54]: Will do

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test and merge

**Comment by tresat** [comment]
@bot-gradle test and merge

**Comment by tresat** [comment]
@bot-gradle test and merge

**Pull request merged**